client/src/lib/dependency-injection/container.js

```diff
@@ -0,0 +1,235 @@
+ * @param middlewareList
+ * @returns {Array}
+ */
+const sortMiddlewares = (middlewareList) => {
+  const graph = [];
+  let sortedMiddlewares = [];
+
+  middlewareList.forEach(entry => {
+    const middleware = normaliseMiddleware(entry);
+    const { name, before, after } = middleware;
+    if (!before.length && !after.length) {
+      after.push('*');
+    }
+    before.forEach(beforeEntry => {
+      graph.push([name, beforeEntry]);
+    });
+    after.forEach(afterEntry => {
+      graph.push([afterEntry, name]);
+    });
+  });
+
+  /** @var Array sortedByName */
+  const sortedByName = toposort(graph);
+  sortedByName.forEach(name => {
+    sortedMiddlewares = sortedMiddlewares.concat(
+      middlewareList.filter(m => m.name === name)
+    );
+  });
+
+  return sortedMiddlewares;
+};
+
+/**
+ * Empties the state and restarts the injector. Should be used
+ * only for testing purposes.
+ */
+const reset = () => {
+  [middlewares, container].forEach(o => {
+    // eslint-disable-next-line no-param-reassign
+    Object.keys(o).forEach(k => delete o[k]);
+  });
+
+  initialised = false;
+};
+
+/**
+ * Applies a middleware function to compose an existing component
+ * with new properties
+ * @param meta An object of metadata
+ * @param key The name of the dependency to customise
+ * @param factory The function that will compose the dependency. Gets passed the
+ * previous state of composition
+ */
+const customise = (meta, key, factory) => {
+  validateMeta(meta);
+  if (!middlewares[key]) middlewares[key] = [];
+  middlewares[key].push({ ...meta, factory });
+};
+
+/**
+ * Register a dependency. This is the initial version of a dependency that will be
+ * passed to the first link in the middleware chain (if any customisations exist)
+ * @param key The name of the dependency to register
+ * @param value The component to register
+ * @param params
+ */
+const register = (key, value, params = {}) => {
+  if (container[key] && params.force !== true) {
+    throw new Error(`
+      Tried to register service ${key} more than once. This practice is discouraged. Consider
+      using Injector.update() to enhance the service rather than override it completely.
+      Otherwise, invoke the register() function with { force: true } as the third argument.
+     `);
+  }
+  container[key] = value;
+};
+
+/**
+ * Gets a dependency
+ * @param key
+ * @returns Component
+ */
+const get = (key) => {
+  if (!initialised) {
+    throw new Error(`
+      Injector.get(): Attempted to access DI layer before it was initialised.
+      Did you forget to invoke Injector.load()?`
+    );
+  }
+  if (!container[key]) {
+    throw new Error(`Injector.get(): Component ${key} does not exist`);
+  }
+  return container[key];
+};
+
+/**
+ * Updates the injector by callback. First parameter should contain
+ * an object with keys for name, and (optional) "before" and "after" declarations
+ * e.x.
+ * Injector.update(
+ *   {
+ *     name: 'my-module',
```

Comment:
I think this should be called `module` rather than `name`, with `beforeModule` and `afterModule` counterparts. It's a bit confusing to use `name` here, I first thought it was the component name.